### PR TITLE
[Perf][Fun-CosyVoice3] Batch inference for Flow decoder

### PR DESCRIPTION
**Pull request by nagisa-kunhah**
## Motivation

Related #1652

## Modifications

- Add `flow_batch.py`, a local adapter for the pinned CosyVoice Flow/CFM implementation (`074ca6dc9e80a2f424f1f74b48bdd7d3fea531cc`). The adapter:
  - packs variable-length prompt and target tokens with explicit masks;
  - preserves per-request prompt mel conditioning and speaker embeddings;
  - runs classifier-free guidance as the required `2 * B` estimator batch;
  - reuses the upstream deterministic noise prefix and returns each unpadded target mel independently.
- Batch compatible requests in the buffered vocoder. Requests are grouped by total mel length into configurable frame buckets before Flow inference, which keeps padding bounded while preserving original response ordering.
- Keep the original `flow.inference()` path for singleton buckets, disabled batching, TensorRT/non-PyTorch estimators, and incompatible future CosyVoice structures. Unsupported structures are reported once at startup rather than failing a request.
- Enable Flow batching by default in the Fun-CosyVoice3 pipeline configuration: `max_batch_size=8`, `max_batch_wait_ms=2`, and `flow_batch_bucket_frames=50`.
- Expose `enable_flow_batch` and `flow_batch_bucket_frames` as vocoder factory arguments and document how to disable batching or tune the bucket size.
- Add structural/unit tests for masking, CFG layout, variable lengths, fallback behavior, scheduler wiring, ordering, and configuration defaults. Add a real-checkpoint GPU invariance test against the pinned upstream Flow path.

### Scope and non-goals

- This PR does **not** batch HiFT. HiFT remains one invocation per mel in this branch. PR #1655 is complementary and should be integrated separately so Flow batching and exact-length HiFT batching can be evaluated both alone and together.
- This PR does **not** change the causal streaming path. The streaming work in PR #1656 has a distinct chunking/causality contract and should not reuse this buffered adapter blindly.
- The adapter is intentionally tied to the documented CosyVoice commit. It does not monkey-patch the CosyVoice checkout.

## Related Issues

Related #1652

Related works: #1655 (HiFT batching), #1656 (causal streaming vocoder).

### Full SeedTTS WER comparison

Full SeedTTS English and Chinese replays were run with the same branch once with Flow batching disabled (serial upstream Flow) and once with the default batching configuration. The service used `max_running_requests=64`; the Flow batch configuration was `max_batch_size=8`, `max_batch_wait_ms=2`, and `flow_batch_bucket_frames=50`.

| Dataset | Serial Flow corpus WER | Batched Flow corpus WER | Delta |
| --- | ---: | ---: | ---: |
| SeedTTS EN (1,088 samples) | 5.2918% | 5.2918% | +0.0000 pp |
| SeedTTS ZH (2,020 samples) | 0.7453% | 0.7476% | +0.0024 pp |

The small ZH difference is well below a material accuracy regression and the EN corpus score is identical for the two runs.

### End-to-end full SeedTTS throughput

The benchmark is an HTTP end-to-end replay on an NVIDIA H20 with Fun-CosyVoice3-0.5B-2512 and the pinned CosyVoice checkout. `main` is included as the repository baseline; the serial-flow row is the same feature branch with the new adapter explicitly disabled, isolating the effect of Flow batching.

| Dataset | `main` QPS | Feature branch, serial Flow QPS | Feature branch, batched Flow QPS | Batched vs. `main` | Batched vs. serial Flow |
| --- | ---: | ---: | ---: | ---: | ---: |
| SeedTTS EN (1,088 requests) | 2.199 | 2.143 | 2.231 | +1.5% | +4.1% |
| SeedTTS ZH (2,020 requests) | 2.099 | 2.069 | 2.238 | +6.6% | +8.2% |

The **e2e benefit is intentionally modest** because Flow is only one part of the critical path. Profiling identified **remaining serial work in HiFT, AR token generation, and reference-audio preprocessing (including CPU ONNX work)**. This is why [HiFT batching in #1655](https://github.com/sgl-project/sglang-omni/pull/1655) is **complementary rather than redundant**: it removes a separate downstream serial bottleneck and makes the Flow-batching benefit more visible end to end.

### Flow-only replay

The standalone Flow benchmark replays the SeedTTS output-length distribution (64 EN + 64 ZH evenly spaced samples), uses 125 synthetic prompt tokens, and models the production 50-frame bucketing rule. These figures measure Flow only and intentionally exclude AR, HiFT, audio I/O, and reference preprocessing.

| Maximum scheduler batch | Flow target tokens/s | Change vs. batch 1 | Actual max Flow batch | Padding overhead |
| --- | ---: | ---: | ---: | ---: |
| 1 | 456.6 | baseline | 1 | 0.00% |
| 2 | 463.8 | +1.6% | 2 | 0.44% |
| 4 | 471.8 | +3.3% | 4 | 0.93% |
| 8 (default) | 484.4 | +6.1% | 6 | 1.63% |
| 16 | 505.2 | +10.6% | 10 | 2.34% |

The configured scheduler batch is an upper bound, not necessarily the actual Flow batch size: differing output lengths split a request batch into separate 50-frame buckets. Batch size 8 is retained as the default because it offers a useful Flow-only gain with low padding and no material e2e latency improvement was observed from increasing it to 16 in the current pipeline.

## Checklist

- [x] Format checked with Ruff for the changed Python files.
- [x] Added unit tests.
- [x] Updated documentation and configuration comments.
- [x] Provided throughput, profiling context, and accuracy results.
- [x] GPU CI has not been requested yet; a maintainer can add `run-ci` and a TTS selector after review.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the `run-ci` label. Once labeled, every subsequent push re-triggers CI as long as the label remains. Use `/tag-and-rerun-ci higgs` or `/tag-and-rerun-ci moss` to select a TTS CI model.

### CODE REVIEW

File: sglang_omni/models/fun_cosyvoice3/stages.py [hunk 1]
```diff
@@ -119,38 +153,144 @@ def prepare_item(
...
+            for index, (state, codes) in enumerate(items)
+        ]
+        results: list[tuple[Any, int] | None] = [None] * len(prepared)
+        serial_requests: list[_PreparedFlowRequest] = []
+        buckets: dict[int, list[_PreparedFlowRequest]] = defaultdict(list)
+        for request in prepared:
+            if self._is_flow_batch_eligible(request.flow_input):
+                buckets[self._flow_bucket_key(request.flow_input)].append(request)
+            else:
+                serial_requests.append(request)
```
Comment (nagisa-kunhah) [inline sglang_omni/models/fun_cosyvoice3/stages.py]: Do we need to keep the serial_requests path here? infer_flow_batch() can also handle a single-item batch, so we could route every valid request through the same Flow implementation and use length bucketing only to control padding.

File: sglang_omni/models/fun_cosyvoice3/flow_batch.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    embedding = F.normalize(packed.embedding, dim=1)
+    embedding = flow.spk_embed_affine_layer(embedding)
+
+    token_embedding = flow.input_embedding(torch.clamp(packed.token, min=0))
+    token_embedding = token_embedding * packed.token_mask.to(token_embedding.dtype)
+    h = flow.pre_lookahead_layer(token_embedding)
+    h = h.repeat_interleave(flow.token_mel_ratio, dim=1)
+    mu = h.transpose(1, 2).contiguous()
+
+    batch_size, channels, max_mel = mu.shape
```
Comment (nagisa-kunhah) [inline sglang_omni/models/fun_cosyvoice3/flow_batch.py]: This may cause OOM, so we need to fix it by adding a `max_batch_size` limit.

**Comment by Kozmosa** [comment]
We re-ran the SeedTTS EN replay (full set, 1,088 samples, voice-clone with ref_text, Qwen3-ASR-1.7B, `max_running_requests=64`, fixed seed) on an H100 across three configurations: `main`, this branch with Flow batching effectively disabled (`--stages.vocoder.factory_args.max_batch_size 1`), and this branch with the default batched config.

| Config | corpus WER |
| --- | ---: |
| `main` (upstream `flow.inference`) | 1.65% |
| serial (B=1 adapter) | 1.52% |
| batched (default config) | 1.52% |

All three land in the same range as the existing Fun-CosyVoice3 reference (1.37%, PR #1331), and batched vs serial shows no accuracy regression. End-to-end EN throughput on our setup is also consistent with the PR's direction (+~1.2% e2e QPS at concurrency 64).

File: sglang_omni/models/fun_cosyvoice3/flow_batch.py
Comment (PoTaTo-Mika) [inline sglang_omni/models/fun_cosyvoice3/flow_batch.py:1]: It is recommended to combine the batch inference feature with the main generation pipeline. This is a code style issue.

**Review by PoTaTo-Mika — APPROVED** [review]
## PR 1663 — CosyVoice3 Perf (aligned with main baseline)

### Setup

| Item | Value |
|---|---|
| Base | `main` @ `470965eb`, PR 1663 stacked on top |
| Model | `FunAudioLLM/Fun-CosyVoice3-0.5B-2512` |
| Dataset | `zhaochenyang20/seed-tts-eval-arrow` |
| Task | Zero-shot voice cloning (`voice_clone=true`, `ref_format=flat`) |
| GPU | 1 × NVIDIA H200 (143 GB) |
| Precision | bf16 |
| Server batch limits | `max_running_requests=64`, `cuda_graph_max_bs=64` |
| Instructions | `null` (no style prompt — same as baseline) |
| Request mode | Non-streaming (`response_format=wav`); TTFA not applicable |
| Samples | en 1088 / zh 2020, full set at every concurrency |
| Warmup | 1 request per concurrency |
| Failures | 0 in all runs |

Baseline and PR 1663 were run with identical config, so the deltas below are apples-to-apples.

### English (en, 1088 samples)

| Concurrency | Success | Failed | QPS (req/s) | Audio throughput (s/s) | Latency mean (s) | Latency median (s) | Latency P95 (s) | Latency P99 (s) | RTF mean | RTF P95 | Audio duration mean (s) | Output tokens/req | Output tokens/s |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| 1 | 1088 | 0 | 1.20 | 5.82 | 0.835 | 0.859 | 1.239 | 1.382 | 0.1840 | 0.2963 | 4.86 | 122 | 146 |
| 2 | 1088 | 0 | 1.86 | 9.01 | 1.072 | 1.041 | 1.569 | 1.768 | 0.2379 | 0.3663 | 4.83 | 121 | 225 |
| 4 | 1088 | 0 | 2.61 | 12.73 | 1.532 | 1.478 | 2.047 | 2.967 | 0.3399 | 0.5328 | 4.88 | 122 | 318 |
| 8 | 1088 | 0 | 2.81 | 13.88 | 2.847 | 2.753 | 3.945 | 5.388 | 0.6393 | 1.0508 | 4.94 | 124 | 347 |
| 16 | 1088 | 0 | 2.87 | 14.28 | 5.555 | 5.428 | 7.150 | 8.462 | 1.2650 | 1.9934 | 4.98 | 124 | 357 |

### Delta vs. main (en)

| Concurrency | main audio thr. (s/s) | PR1663 audio thr. (s/s) | Δ | main QPS | PR1663 QPS | Δ | main latency mean (s) | PR1663 latency mean (s) | Δ | main RTF | PR1663 RTF | Δ |
|---|---|---|---|---|---|---|---|---|---|---|---|---|
| 1 | 5.84 | 5.82 | -0.3% | 1.23 | 1.20 | -2.3% | 0.816 | 0.835 | +2.3% | 0.1817 | 0.1840 | +1.3% |
| 2 | 9.09 | 9.01 | -1.0% | 1.89 | 1.86 | -1.4% | 1.057 | 1.072 | +1.4% | 0.2328 | 0.2379 | +2.2% |
| 4 | 12.56 | 12.73 | +1.4% | 2.54 | 2.61 | +2.8% | 1.576 | 1.532 | -2.8% | 0.3465 | 0.3399 | -1.9% |
| 8 | 13.37 | 13.88 | +3.8% | 2.68 | 2.81 | +4.7% | 2.978 | 2.847 | -4.4% | 0.6657 | 0.6393 | -4.0% |
| 16 | 13.53 | 14.28 | +5.5% | 2.78 | 2.87 | +3.4% | 5.747 | 5.555 | -3.3% | 1.3105 | 1.2650 | -3.5% |

### Chinese (zh, 2020 samples)

| Concurrency | Success | Failed | QPS (req/s) | Audio throughput (s/s) | Latency mean (s) | Latency median (s) | Latency P95 (s) | Latency P99 (s) | RTF mean | RTF P95 | Audio duration mean (s) | Output tokens/req | Output tokens/s |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| 1 | 2020 | 0 | 1.09 | 7.14 | 0.914 | 0.926 | 1.305 | 1.490 | 0.1426 | 0.2005 | 6.53 | 163 | 178 |
| 2 | 2020 | 0 | 1.70 | 11.03 | 1.175 | 1.157 | 1.551 | 1.739 | 0.1840 | 0.2310 | 6.48 | 162 | 276 |
| 4 | 2020 | 0 | 2.35 | 15.52 | 1.699 | 1.657 | 2.181 | 2.909 | 0.2645 | 0.3464 | 6.59 | 165 | 388 |
| 8 | 2020 | 0 | 2.89 | 18.86 | 2.763 | 2.693 | 3.607 | 4.610 | 0.4361 | 0.5864 | 6.52 | 163 | 472 |
| 16 | 2020 | 0 | 2.86 | 18.72 | 5.585 | 5.487 | 6.890 | 7.569 | 0.8902 | 1.1944 | 6.55 | 164 | 468 |

### Delta vs. main (zh)

| Concurrency | main audio thr. (s/s) | PR1663 audio thr. (s/s) | Δ | main QPS | PR1663 QPS | Δ | main latency mean (s) | PR1663 latency mean (s) | Δ | main RTF | PR1663 RTF | Δ |
|---|---|---|---|---|---|---|---|---|---|---|---|---|
| 1 | 7.21 | 7.14 | -1.0% | 1.10 | 1.09 | -0.9% | 0.905 | 0.914 | +1.0% | 0.1412 | 0.1426 | +1.0% |
| 2 | 11.11 | 11.03 | -0.8% | 1.72 | 1.70 | -1.3% | 1.160 | 1.175 | +1.3% | 0.1819 | 0.1840 | +1.2% |
| 4 | 15.54 | 15.52 | -0.1% | 2.40 | 2.35 | -1.9% | 1.667 | 1.699 | +1.9% | 0.2625 | 0.2645 | +0.8% |
| 8 | 17.68 | 18.86 | +6.7% | 2.76 | 2.89 | +4.7% | 2.892 | 2.763 | -4.5% | 0.4625 | 0.4361 | -5.7% |
| 16 | 17.81 | 18.72 | +5.1% | 2.70 | 2.86 | +5.8% | 5.868 | 5.585 | -4.8% | 0.9339 | 0.8902 | -4.7% |

### Takeaway

- **c=8 / c=16: consistent win.** Audio throughput +3.8% / +5.5% (en) and +6.7% / +5.1% (zh), with latency and RTF moving down 3–6% in the same direction across both languages.
- **c≤4: flat.** All deltas sit within ±3%, i.e. no measurable regression and no measurable gain.
- **Shape matches the change.** Gains grow monotonically with concurrency, which is the expected signature of a large-batch (flow batch) optimization, and argues against the c=8/16 deltas being run-to-run noise.
- **Caveat:** one run per config, so the noise band (~±3%, inferred from the c≤4 spread) is not measured directly. The direction is solid; treat the exact percentages as approximate.

File: sglang_omni/models/fun_cosyvoice3/stages.py [hunk 3]
```diff
@@ -24,7 +26,10 @@
 from sglang_omni.utils.checkpoint import resolve_checkpoint
 from sglang_omni.utils.device import resolve_device_spec
 
-logger = logging.getLogger(__name__)
+# This is an admission budget, not a maximum supported request length. The
```
Comment (SandyLuXY) [inline sglang_omni/models/fun_cosyvoice3/stages.py]: please follow comment style `# Note (xxx): `

File: tests/unit_test/fun_cosyvoice3/test_flow_batch.py
Comment (SandyLuXY) [inline tests/unit_test/fun_cosyvoice3/test_flow_batch.py:1]: update `tests/README.md` for cosy voice tests

File: sglang_omni/models/fun_cosyvoice3/stages.py (hunk 3 above)
Comment (nagisa-kunhah) [reply sglang_omni/models/fun_cosyvoice3/stages.py]: Thanks, fix it.

File: tests/unit_test/fun_cosyvoice3/test_flow_batch.py
Comment (nagisa-kunhah) [reply tests/unit_test/fun_cosyvoice3/test_flow_batch.py:1]: Done.

**Review by SandyLuXY — APPROVED** [review]
